sling/nlp/parser/tools/train_pytorch.py

```diff
@@ -0,0 +1,377 @@
+        self.moving_avg_coeff = args.moving_average_coeff
+
+
+    # Returns a string representation of all hyperparameters.
+    def __str__(self):
+      return str(self.__dict__)
+
+
+  # Instantiates the trainer with the given model, optional evaluator,
+  # and hyperparameters.
+  def __init__(self, sempar, evaluator=None, output_file_prefix=None, \
+               hyperparams=Hyperparams()):
+    self.model = sempar
+    self.evaluator = evaluator
+    self.hyperparams = hyperparams
+
+    if hyperparams.optimizer == "sgd":
+      self.optimizer = torch.optim.SGD(
+        sempar.parameters(),
+        lr=self.hyperparams.lr,
+        momentum=0,
+        dampening=0,
+        weight_decay=0,
+        nesterov=False)
+    elif hyperparams.optimizer == "adam":
+      self.optimizer = torch.optim.Adam(
+          sempar.parameters(), lr=hyperparams.lr, weight_decay=0, \
+              betas=(hyperparams.adam_beta1, hyperparams.adam_beta2), \
+              eps=hyperparams.adam_eps)
+
+    num_params = 0
+    for name, p in sempar.named_parameters():
+      if p.requires_grad:
+        print name, ":", p.size()
+        num_params += torch.numel(p)
+    print "Number of parameters:", num_params
+
+    self.count = 0
+    self.last_eval_count = 0
+
+    self.current_batch_num_transitions = 0
+    self.current_batch_size = 0
+    self.batch_loss = Var(torch.FloatTensor([0.0]))
+    self._reset()
+
+    self.checkpoint_metrics = []
+    self.best_metric = None
+    self.output_file_prefix = output_file_prefix
+
+    # Exponential moving average clones.
+    self.averages = {}
+    if hyperparams.moving_avg:
+      for name, p in sempar.named_parameters():
+        if p.requires_grad:
+          self.averages[name] = p.data.clone()
+
+
+  # Resets the state for a new batch.
+  def _reset(self):
+    self.current_batch_size = 0
+    self.current_batch_num_transitions = 0
+    del self.batch_loss
+    self.batch_loss = Var(torch.FloatTensor([0.0]))
+    self.optimizer.zero_grad()
+
+
+  # Processes a single given example.
+  def process(self, example):
+    loss, num_transitions = self.model.forward(example, train=True)
+    self.current_batch_num_transitions += num_transitions
+    self.batch_loss += loss
+    self.current_batch_size += 1
+    self.count += 1
+    if self.current_batch_size == self.hyperparams.batch_size:
+      self.update()
+    if self.count % self.hyperparams.report_every == 0:
+      self.evaluate()
+
+
+  # Clips the gradients separately for each PyTorch Parameter.
+  def clip_gradients(self):
+    if self.hyperparams.gradient_clip is not None:
+      for p in self.model.parameters():
+        torch.nn.utils.clip_grad_norm([p], self.hyperparams.gradient_clip)
+
+
+  # Performs a gradient update.
+  def update(self):
+    if self.current_batch_size > 0:
+      start = time.time()
+      self.batch_loss /= self.current_batch_num_transitions
+
+      # Add the regularization penalty to the batch loss.
```

Comment:
I did not know that we used L2 regularization. Does it actually help?

Reply:
We have always used L2=1e-4. This was the default value in DRAGNN's hyperparameter protobuf, which is perhaps why you didn't explicitly see it.  During our initial grid search this was better than no L2 regularization.